doc/rewriting-services.md

```diff
@@ -0,0 +1,204 @@
+# Tips for rewriting legacy services
+
+## First, write some tests
+
+If service tests don’t exist for the legacy service, stop and write them first.
+It’s recommended to PR these separately. If there’s some test coverage, it’s
+probably fine to move right ahead and add more in the process. Make sure the
+tests are passing, though.
+
+## Organization
+
+1. When there’s a single legacy service that handles lots of different things
+(e.g. version, license, and downloads), it should be split into three separate
+service classes and placed in three separate files.
+`myservice-version.service.js`, `myservice-license.service.js`,
+`myservice-downloads.service.js`.
+
+2. When a service has different variants of basically the same thing, it’s okay
+to put them in the same service. For example, daily/weekly/monthly/total
+downloads can go in one badge, and star rating vs point rating vs rating count
+can go in one badge, and same with various kinds of detail about a pull request.
+The hard limit (as of now anyway) is *one category per service*.
+
+3. If the tests haven’t been split up, split them up too and make sure they
+still pass.
+
+## Get the route working
+
+1. Disable the legacy service by adding a `return` at the top of
+`registerLegacyRouteHandler()`.
+
+2. Set up the route for one of the badges. First determine if you can express
+the route using a `pattern`. A `pattern` is the simplest way to declare the
+route, also the most readable, and will be useful for displaying a badge builder
+with fields in the front end and generating badge URLs programmatically. One
+limitation to keep in mind is that, at present, the trailing parameter of a
+pattern can't be optional. If the last part of a route is optional, like a
+branch, you will need to use a `format` regex.
```

Comment:
It would be nice to add a bit more explanation of the pattern/format thing here. Perhaps the minimum is just an example. e.g:

> A `pattern` (e.g: `pattern: ':param1/:param2',`) is the simplest way...

> ...you will need to use a format regex. (e.g: `format: '([^/]+/[^/]+)'` )

but maybe this whole topic of defining routes needs its own more detailed "chapter" of docs we can just link to here.

Reply:
Yea, one thing that has occurred to me while writing this is that once the rewrite is done, it could serve as the basis of a developer guide to services – something that goes into more depth than the tutorial, and also talks about many of the various edge cases and contingencies.